[EOBS-1225] Food & Fluid empty periods are recorded

homostellaris opened the pull request:
# Summary
A new section in the reports entitled "Food and Fluid Period With No Entries" lists food and fluid periods during which no food and fluid observations were performed.

These 'empty' periods are only shown for periods since the period in which the first food and fluid observation was ever performed for a particular spell. Periods before the first food and fluid observation do not count.

Contiguous 'empty' periods are displayed in a single entry, with the displayed _start time_ being the start of the first empty periods and the displayed _end time_ being the end of the last empty period. A string of empty periods is broken by either a period that has observations or the currently active period.

# Implementation
A `get_empty_period` method has been added to the `nh_clinical_patient_observation_food_fluid` model. It takes the previously generated _period dictionaries_ as an input and determines what the empty periods are based on those by looking at the time gaps between them. It creates the entry for the report by taking the end datetime of one 'normal' period and the start datetime of the next 'normal' period chronologically.

homostellaris committed:
Get empty periods method added.

homostellaris committed:
Various fixes.

Fixed `__init__`.
Added empty periods to report data.
Also changed the start and end of the range used in the `get_empty_periods` method to better reflect the variable name which is `all_period_start_days` (before it was excluding the start and end periods).
Fixed flake8 errors.

homostellaris committed:
Added empty periods to report.

homostellaris committed:
Fixed an error in the implementation.

Contiguous empty periods only create one entry with start datetime from the first empty period and the end datetime from the last empty period, rather than an entry for every period.
Added more tests.

homostellaris committed:
Fixed unintended side effect.

`get_empty_periods` method was appending the a dictionary to represent the active period to the passed list of other period dictionaries. There was not meant to be any side effects as the extra dictionary is only needed for logic within the method.

 Also fixed some tests.

homostellaris committed:
Fixed flake8 error.

homostellaris committed:
Merge remote-tracking branch 'origin/develop' into EOBS-1225-food-and-fluid-empty-periods-are-recorded

homostellaris committed:
Added some docstrings.

moderatemisbehaviour commented on nh_food_and_fluid/views/observation_report_template.xml:
```diff
@@ -56,5 +61,23 @@
+                </table>
+            </div>
+        </template>
+
```
@Gimpneek suggested adding column headings for extra clarity, I agree.

Gimpneek replied:
Let's add some column headings on the table to make things easier to read

homostellaris committed:
Added headers and top margin.

Headers added for clarity.
Top margin added by copying classes from other sections.

homostellaris committed:
Merge remote-tracking branch 'origin/develop' into EOBS-1225-food-and-fluid-empty-periods-are-recorded

technotariat committed:
Merge branch 'develop' into EOBS-1225-food-and-fluid-empty-periods-are-recorded

technotariat committed:
- Before moving on 1327

homostellaris commented:
Closing because I can't find another way to remove it from my _pull requests_ view!
